modmail/dispatcher.py

```diff
@@ -0,0 +1,167 @@
+
+        Priority is optional. If provided, the handler will be called in order according to it's priority.
+        If the handler returns True, the event is considered "handled" and further dispatch is cancelled.
+        Lower numbers go first. Negative numbers are supported.
+
+        If priority is not provided the event is dispatched asynchronously after all blocking handlers,
+         and the return result of the handler has no effect.
+        """
+        if func:
+            self._register_handler(event_name, priority, func)
+            return func
+
+        def register_decorator(func: CoroutineFunction) -> CoroutineFunction:
+            self._register_handler(event_name, priority, func)
+            return func
+
+        return register_decorator
+
+    def unregister(self, func: CoroutineFunction, *event_names: Optional[str]) -> None:
+        """
+        Unregister a function from dispatch.
+
+        If event_names is not provided, removes it from all handlers.
+        If you have accidentally added a handler twice to an event, this will only remove one instance.
+        """
+        if not event_names:
+            event_names = self.handlers.keys()
+
+        for event_name in event_names:
+            if event_name not in self.handlers:
+                logger.exception(
+                    "Attempted to unregister handler %r from event name %s, which wasn't registered.",
+                    func,
+                    event_name,
+                )
+                continue
+
+            if func in self.handlers[event_name]:
+                self.handlers[event_name].remove(func)
+
+            if func in self.blocking_handlers[event_name]:
+                # blocking handlers are two separate lists because it makes searching for items
+                # (and thus bisect) a hundred times easier. But that does mean we have to keep them in sync.
+                index = self.blocking_handlers[event_name].index(func)
+                del self.blocking_handlers[event_name][index]
+                del self.blocking_priorities[event_name][index]
+
+    async def dispatch(self, event_name: str, *args, **kwargs) -> None:
+        """
+        Trigger dispatch of an event, passing args directly to each handler.
+
+        Beware passing mutable args--previous handlers, if misbehaving, can mutate them.
+        """
+        if event_name not in self.blocking_handlers:
+            logger.exception(
+                "Unregistered event '%s' was dispatched to no handlers with data: %r %r",
+                event_name,
+                args,
+                kwargs,
+            )
+            self.blocking_handlers[event_name] = []
+            self.handlers[event_name] = []
+            self.blocking_priorities[event_name] = []
+
+        for handler in self.blocking_handlers[event_name]:
+            if await handler(*args, **kwargs):
+                return
```

Comment:
Why does it return here with None? If its a blocking handler, it should only be blocked if the awaitable returns a Nonzero value.

This function should return the error code so the caller *knows* if the checks failed or not.

Reply:
We return none here in order to cancel further dispatch of the vent